lib/rucio/core/oidc.py

```diff
@@ -0,0 +1,925 @@
+    :param ip: IP address of the client as a string.
+    :param session: The database session in use.
+
+    :returns: One of the following tuples: ("fetchcode", <code>); ("token", <token>);
+              ("polling", True); The result depends on the authentication strategy being used
+              (no auto, auto, polling).
+    """
+    try:
+        start = time.time()
+        parsed_authquery = parse_qs(auth_query_string)
+        state = parsed_authquery["state"][0]
+        code = parsed_authquery["code"][0]
+        # getting oauth request params from the oauth_requests DB Table
+        oauth_req_params = session.query(models.OAuthRequest).filter_by(state=state).first()
+        if oauth_req_params is None:
+            raise CannotAuthenticate("User related Rucio OIDC session could not keep "
+                                     + "track of responses from outstanding requests.")  # NOQA: W503
+        req_url = urlparse(oauth_req_params.redirect_msg)
+        issuer = 'https://' + req_url.netloc + '/'
+        req_params = parse_qs(req_url.query)
+        for key in req_params:
+            req_params[key] = val_to_space_sep_str(req_params[key])
+
+        oidc_client = __get_init_oidc_client(issuer=issuer, code=code, **req_params)['client']
+        record_counter(counters='IdP_authentication.code_granted')
+        # exchange access code for a access token
+        oidc_tokens = oidc_client.do_access_token_request(state=state,
+                                                          request_args={"code": code},
+                                                          authn_method="client_secret_basic")
+        if 'error' in oidc_tokens:
+            raise CannotAuthorize(oidc_tokens['error'])
+        # mitigate replay attacks
+        nonce = oauth_req_params.nonce
+        if oidc_tokens['id_token']['nonce'] != nonce:
+            raise CannotAuthenticate("ID token could not be associated with the Rucio OIDC Client"
+                                     + " session. This points to possible replay attack !")  # NOQA: W503
+
+        # starting to fill dictionary with parameters for token DB row
+        jwt_row_dict, extra_dict = {}, {}
+        jwt_row_dict['identity'] = oidc_identity_string(oidc_tokens['id_token']['sub'],
+                                                        oidc_tokens['id_token']['iss'])
+        jwt_row_dict['account'] = oauth_req_params.account
+
+        if jwt_row_dict['account'] == InternalAccount('webui'):
+            try:
+                jwt_row_dict['account'] = get_default_account(jwt_row_dict['identity'], IdentityType.OIDC, True, session=session)
+            except Exception:
+                return {'webhome': None, 'token': None}
+
+        # check if given account has the identity registered
+        if not exist_identity_account(jwt_row_dict['identity'], IdentityType.OIDC, jwt_row_dict['account']):
```

Comment:
This is also a session leak here, you have to hand through the session.
I did not check all the method calls, but in general please verify again that if you call from a method, where you use the db session another method, which uses the session, you have to hand it through, otherwise multiple sessions will be started and this get's transactionally inconsistent.